### PR TITLE
En 9502/validator to delegation

### PR DESCRIPTION
**Pull request by sasurobert**
The first part of the implementation of moving validators to delegation contracts. All being optional.
Implemented the first version where the owner makes a new contract and moves the data there.

### CODE REVIEW

File: vm/systemSmartContracts/delegationManager.go [hunk 1]
```diff
@@ -240,6 +259,50 @@ func (d *delegationManager) createNewDelegationContract(args *vmcommon.ContractC
...
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) != 2 {
+		d.eei.AddReturnMessage("invalid number of arguments")
+		return vmcommon.UserError
+	}
+
+	arguments := append([][]byte{args.CallerAddr}, args.Arguments...)
+
+	return d.deployNewContract(args, false, "initFromValidatorData", arguments)
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/delegationManager.go:283]: maybe define a const for the `initFromValidatorData` ?

File: vm/systemSmartContracts/validator.go [hunk 2]
```diff
@@ -1828,6 +1836,110 @@ func (v *validatorSC) getTotalStakedTopUpStakedBlsKeys(args *vmcommon.ContractCa
...
+	err = v.saveRegistrationData(newAddress, validatorData)
+	if err != nil {
+		v.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (v *validatorSC) changeRewardAndOwnerAddressOnStaking(registrationData *ValidatorDataV2) vmcommon.ReturnCode {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/validator.go:1924]: changeOwnerAndRewardAddressOnStaking ? as to be consistent on naming

File: vm/systemSmartContracts/staking.go [hunk 3]
```diff
@@ -1706,6 +1712,54 @@ func (s *stakingSC) cleanAdditionalQueue(args *vmcommon.ContractCallInput) vmcom
...
+		s.eei.AddReturnMessage("CallValue must be 0")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) < 2 {
+		s.eei.AddReturnMessage("number of arguments is 2 at minimum")
+		return vmcommon.UserError
+	}
+
+	newOwnerAddress := args.Arguments[0]
+	if !core.IsSmartContractAddress(newOwnerAddress) {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/staking.go:1733]: IsSmartContractOnMetachain ?

File: vm/systemSmartContracts/staking.go [hunk 4]
```diff
@@ -1706,6 +1712,54 @@ func (s *stakingSC) cleanAdditionalQueue(args *vmcommon.ContractCallInput) vmcom
...
+	for i := 1; i < len(args.Arguments); i++ {
+		blsKey := args.Arguments[i]
+		registrationData, err := s.getOrCreateRegisteredData(blsKey)
+		if err != nil {
+			s.eei.AddReturnMessage(err.Error())
+			return vmcommon.UserError
+		}
+
+		if len(registrationData.RewardAddress) == 0 {
+			s.eei.AddReturnMessage("cannot unStake a key that is not registered")
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/staking.go:1747]: cannot changeOwnerAndRewardAddress a key that is not registered *

File: vm/systemSmartContracts/validator.go [hunk 5]
```diff
@@ -216,6 +220,10 @@ func (v *validatorSC) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnC
 		return v.getUnStakedTokensList(args)
 	case "reStakeUnStakedNodes":
 		return v.reStakeUnStakedNodes(args)
+	case "mergeValidatorData":
+		return v.mergeValidatorData(args)
+	case "changeOwnerOfValidatorData":
+		return v.changeOwnerOfValidatorData(args)
 	}
 
 	v.eei.AddReturnMessage("invalid method to call")
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/validator.go:229]: 3 system SCs with 3 different return messages on an undefined functions :(

File: vm/systemSmartContracts/validator.go [hunk 6]
```diff
@@ -1828,6 +1836,110 @@ func (v *validatorSC) getTotalStakedTopUpStakedBlsKeys(args *vmcommon.ContractCa
...
+	if !bytes.Equal(args.CallerAddr, v.delegationMgrSCAddress) {
+		v.eei.AddReturnMessage("invalid caller address")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		v.eei.AddReturnMessage("callValue must be 0")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) != 2 {
+		v.eei.AddReturnMessage("invalid number of arguments, needed 2")
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/validator.go:1853]: we have different messages for the same error. Let's make them uniform, at least for the new functions. Check delegationManager.go L277

File: vm/systemSmartContracts/validator.go [hunk 7]
```diff
@@ -1828,6 +1836,110 @@ func (v *validatorSC) getTotalStakedTopUpStakedBlsKeys(args *vmcommon.ContractCa
...
+		return vmcommon.UserError
+	}
+	if len(args.Arguments[0]) != len(args.CallerAddr) || len(args.Arguments[1]) != len(args.CallerAddr) {
+		v.eei.AddReturnMessage("invalid argument, wanted an address")
+		return vmcommon.UserError
+	}
+	if !core.IsSmartContractAddress(args.Arguments[1]) {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/validator.go:1860]: IsSmartContractOnMetachain ?

File: vm/systemSmartContracts/validator.go [hunk 8]
```diff
@@ -1828,6 +1836,110 @@ func (v *validatorSC) getTotalStakedTopUpStakedBlsKeys(args *vmcommon.ContractCa
...
+		v.eei.AddReturnMessage("destination address must be a delegation smart contract")
+		return vmcommon.UserError
+	}
+	if bytes.Equal(args.Arguments[0], args.Arguments[1]) {
+		v.eei.AddReturnMessage("sender and destination addresses are equal")
+		return vmcommon.UserError
+	}
+	if core.IsSmartContractAddress(args.Arguments[0]) {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/validator.go:1868]: why? Maybe someone had a multisig contract as a validator and now wants to use the builtin delegation system.

File: vm/systemSmartContracts/validator.go [hunk 9]
```diff
@@ -1828,6 +1836,110 @@ func (v *validatorSC) getTotalStakedTopUpStakedBlsKeys(args *vmcommon.ContractCa
...
+	if args.CallValue.Cmp(zero) != 0 {
+		v.eei.AddReturnMessage("callValue must be 0")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) != 2 {
+		v.eei.AddReturnMessage("invalid number of arguments, needed 2")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments[0]) != len(args.CallerAddr) || len(args.Arguments[1]) != len(args.CallerAddr) {
+		v.eei.AddReturnMessage("invalid argument, wanted an address")
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/validator.go:1857]: invalid arguments, wanted an address for the first and the second arguments

File: vm/systemSmartContracts/validator.go [hunk 10]
```diff
@@ -1828,6 +1836,110 @@ func (v *validatorSC) getTotalStakedTopUpStakedBlsKeys(args *vmcommon.ContractCa
...
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		v.eei.AddReturnMessage("callValue must be 0")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) != 2 {
+		v.eei.AddReturnMessage("invalid number of arguments, needed 2")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments[0]) != len(args.CallerAddr) || len(args.Arguments[1]) != len(args.CallerAddr) {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/validator.go:1856]: the code would be easier to follow if you used:
```GO
oldAddress := args.Arguments[0]
newAddress := args.Arguments[1]
```

same as in the `changeOwnerOfValidatorData` function

File: vm/systemSmartContracts/validator.go [hunk 11]
```diff
@@ -1828,6 +1836,110 @@ func (v *validatorSC) getTotalStakedTopUpStakedBlsKeys(args *vmcommon.ContractCa
...
+	if err != nil {
+		v.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (v *validatorSC) changeRewardAndOwnerAddressOnStaking(registrationData *ValidatorDataV2) vmcommon.ReturnCode {
+	txData := "changeOwnerAndRewardAddress@" + hex.EncodeToString(registrationData.RewardAddress)
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/validator.go:1925]: is this implemented on the staking SC?

File: vm/systemSmartContracts/validator.go [hunk 12]
```diff
@@ -1828,6 +1836,110 @@ func (v *validatorSC) getTotalStakedTopUpStakedBlsKeys(args *vmcommon.ContractCa
 	return vmcommon.Ok
 }
 
+func (v *validatorSC) checkInputArgsForValidatorToDelegation(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/validator.go:1839]: unit tests for the new code?

**Review by iulianpascalau — CHANGES_REQUESTED** [review]
Maybe we can find a way to deal with the duplicated code on the loading of the staking/validator data as the future changes will work on all system SC we have.

File: vm/systemSmartContracts/delegationManager.go [hunk 13]
```diff
@@ -222,7 +238,10 @@ func (d *delegationManager) createNewDelegationContract(args *vmcommon.ContractC
 	delegationManagement.LastAddress = newAddress
 	delegationList.Addresses = append(delegationList.Addresses, newAddress)
 
-	d.eei.SetStorage(args.CallerAddr, newAddress)
+	currentStorage := d.eei.GetStorage(args.CallerAddr)
+	currentStorage = append(currentStorage, newAddress...)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegationManager.go:242]: yes. it is a list of addresses.

File: vm/systemSmartContracts/staking.go (hunk 3 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:1733]: no need as this function has check who can call it.

File: vm/systemSmartContracts/validator.go (hunk 5 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go:229]: that's it. development in progress. :|

File: vm/systemSmartContracts/validator.go (hunk 7 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go:1860]: it is harder to use that function. you need the metachain identifier as well. This should be enough as it is protected by caller

File: vm/systemSmartContracts/validator.go (hunk 8 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go:1868]: Because that is not the current case. It is better like this as you do not know how the underlying contract works. The only SC is our RUST legacy.

File: vm/systemSmartContracts/validator.go (hunk 11 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go:1925]: yes

File: vm/systemSmartContracts/validator.go (hunk 12 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go:1839]: in some subsequent PR.

File: vm/systemSmartContracts/staking.go (hunk 4 above)
Comment (SebastianMarian) [reply vm/systemSmartContracts/staking.go:1747]: :+1:

File: vm/systemSmartContracts/validator.go [hunk 14]
```diff
@@ -1828,6 +1836,110 @@ func (v *validatorSC) getTotalStakedTopUpStakedBlsKeys(args *vmcommon.ContractCa
...
+	oldAddress := args.Arguments[0]
+	newAddress := args.Arguments[1]
+
+	validatorData, err := v.getOrCreateRegistrationData(oldAddress)
+	if err != nil {
+		v.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if len(validatorData.RewardAddress) == 0 {
+		v.eei.AddReturnMessage("validator data do not exists")
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/validator.go:1900]: does

File: vm/systemSmartContracts/staking.go [hunk 15]
```diff
@@ -1706,6 +1712,54 @@ func (s *stakingSC) cleanAdditionalQueue(args *vmcommon.ContractCallInput) vmcom
...
+func (s *stakingSC) changeOwnerAndRewardAddress(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !s.flagValidatorToDelegation.IsSet() {
+		return vmcommon.UserError
+	}
+	if !bytes.Equal(args.CallerAddr, s.stakeAccessAddr) {
+		s.eei.AddReturnMessage("change owner and reward address can be called by validator SC only")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		s.eei.AddReturnMessage("CallValue must be 0")
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/staking.go:1724]: callValue

File: vm/mock/systemEIStub.go [hunk 16]
```diff
@@ -114,13 +114,7 @@ func (s *SystemEIStub) ExecuteOnDestContext(
...
 
 // DeploySystemSC -
-func (s *SystemEIStub) DeploySystemSC(
-	baseContract []byte,
-	newAddress []byte,
-	ownerAddress []byte,
-	value *big.Int,
-	input [][]byte,
-) (vmcommon.ReturnCode, error) {
+func (s *SystemEIStub) DeploySystemSC(baseContract []byte, newAddress []byte, ownerAddress []byte, initFunction string, value *big.Int, input [][]byte) (vmcommon.ReturnCode, error) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/mock/systemEIStub.go:117]: done

File: vm/systemSmartContracts/delegation.go [hunk 17]
```diff
@@ -301,7 +324,165 @@ func (d *delegation) init(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
...
+	}
+	if len(args.Arguments[0]) != len(d.delegationMgrSCAddress) {
+		d.eei.AddReturnMessage("invalid arguments, first must be an address")
+		return vmcommon.UserError
+	}
+	return vmcommon.Ok
+}
+
+func (d *delegation) initFromValidatorData(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	returnCode := d.checkArgumentsForGeneralViewFunc(args)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:355]: done

File: process/scToProtocol/stakingToPeer.go [hunk 18]
```diff
@@ -295,6 +298,14 @@ func (stp *stakingToPeer) updatePeerState(
 		}
 		account.SetUnStakedEpoch(stakingData.UnStakedEpoch)
 
+		if stp.flagValidatorToDelegation.IsSet() && !bytes.Equal(account.GetRewardAddress(), stakingData.RewardAddress) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/scToProtocol/stakingToPeer.go:301]: it is one if only.  with a setreward address. it has to be called for existing accounts only - this is an edge case. but must be treated. unit tests in subsequent PRs.

File: vm/systemSmartContracts/delegation.go [hunk 19]
```diff
@@ -301,7 +324,165 @@ func (d *delegation) init(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
...
+		d.eei.AddReturnMessage("clean unStaked info before changing validator to delegation contract")
+		return vmcommon.UserError
+	}
+
+	returnCode = d.initDelegationStructures(delegationManagement.MinDeposit, ownerAddress, serviceFee, maxDelegationCap)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	dStatus := &DelegationContractStatus{
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:412]: this is only an initialization. made a func for init.

[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:412]: moved to new function

File: vm/systemSmartContracts/delegation.go [hunk 20]
```diff
@@ -301,7 +324,165 @@ func (d *delegation) init(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
...
+
+	return validatorData, nil
+}
+
+func (d *delegation) mergeValidatorDataToDelegation(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !d.flagValidatorToDelegation.IsSet() {
+		d.eei.AddReturnMessage(args.Function + " is an unknown function")
+		return vmcommon.UserError
+	}
+	return vmcommon.Ok
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:485]: exactly. that is a little more complicated.

File: vm/systemSmartContracts/delegation.go [hunk 21]
```diff
@@ -301,7 +324,165 @@ func (d *delegation) init(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
 		return vmcommon.UserError
 	}
 
-	return d.delegateUser(initialOwnerFunds, ownerAddress, args.RecipientAddr, dStatus)
+	return vmcommon.Ok
+}
+
+func (d *delegation) checkArgumentsForValidatorToDelegation(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:330]: next PR

File: vm/systemSmartContracts/delegation.go [hunk 22]
```diff
@@ -2461,6 +2642,9 @@ func (d *delegation) EpochConfirmed(epoch uint32) {
 
 	d.stakingV2Enabled.Toggle(epoch > d.stakingV2EnableEpoch)
 	log.Debug("stakingV2", "enabled", d.stakingV2Enabled.IsSet())
+
+	d.flagValidatorToDelegation.Toggle(epoch > +d.validatorToDelegationEnableEpoch)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:2646]: done

File: vm/systemSmartContracts/delegationManager.go (hunk 1 above)
Comment (sasurobert) [reply vm/systemSmartContracts/delegationManager.go:283]: done

File: vm/systemSmartContracts/validator.go (hunk 2 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go:1924]: done

File: vm/systemSmartContracts/staking.go (hunk 4 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:1747]: done

File: vm/systemSmartContracts/validator.go (hunk 6 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go:1853]: every function is done differently.

File: vm/systemSmartContracts/validator.go (hunk 9 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go:1857]: done

File: vm/systemSmartContracts/validator.go (hunk 10 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go:1856]: done

File: cmd/node/config/systemSmartContractsConfig.toml [hunk 23]
```diff
@@ -37,5 +37,6 @@
 
 [DelegationSystemSCConfig]
     EnabledEpoch   = 1 #enable epoch should not be 0
+    ValidatorToDelegatorEnableEpoch = 3
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/config/systemSmartContractsConfig.toml:40]: done

File: config/systemSmartContractsConfig.go [hunk 24]
```diff
@@ -48,10 +48,11 @@ type GovernanceSystemSCConfig struct {
 
 // DelegationManagerSystemSCConfig defines a set of constants to initialize the delegation manager system smart contract
 type DelegationManagerSystemSCConfig struct {
-	MinCreationDeposit  string
-	EnabledEpoch        uint32
-	MinStakeAmount      string
-	ConfigChangeAddress string
+	MinCreationDeposit              string
+	EnabledEpoch                    uint32
+	ValidatorToDelegatorEnableEpoch uint32
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply config/systemSmartContractsConfig.go:53]: done

File: vm/systemSmartContracts/delegation.go [hunk 25]
```diff
@@ -301,7 +324,165 @@ func (d *delegation) init(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
...
+		nodesData := &NodesData{
+			BLSKey:    blsKey,
+			SignedMsg: blsKey,
+		}
+		switch status {
+		case active:
+			dStatus.StakedKeys = append(dStatus.StakedKeys, nodesData)
+		case unStaked:
+			dStatus.UnStakedKeys = append(dStatus.StakedKeys, nodesData)
+		}
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:434]: they will not be populated.

File: vm/systemSmartContracts/delegation.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:2646]: done

File: vm/systemSmartContracts/delegationManager.go [hunk 26]
```diff
@@ -240,6 +259,50 @@ func (d *delegationManager) createNewDelegationContract(args *vmcommon.ContractC
...
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		d.eei.AddReturnMessage("callValue must be 0")
+		return vmcommon.UserError
+	}
+	err := d.eei.UseGas(d.gasCost.MetaChainSystemSCsCost.ValidatorToDelegation)
+	if err != nil {
+		d.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegationManager.go:274]: done

File: vm/systemSmartContracts/delegationManager.go [hunk 27]
```diff
@@ -240,6 +259,50 @@ func (d *delegationManager) createNewDelegationContract(args *vmcommon.ContractC
...
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegationManager.go:298]: done

File: vm/systemSmartContracts/delegationManager.go [hunk 28]
```diff
@@ -430,6 +493,9 @@ func (d *delegationManager) SetNewGasCost(gasCost vm.GasCost) {
 func (d *delegationManager) EpochConfirmed(epoch uint32) {
 	d.delegationMgrEnabled.Toggle(epoch >= d.enableDelegationMgrEpoch)
 	log.Debug("delegationManager", "enabled", d.delegationMgrEnabled.IsSet())
+
+	d.flagValidatorToDelegation.Toggle(epoch >= d.validatorToDelegationEnableEpoch)
+	log.Debug("delegationManager validator to delegation", "enabled", d.flagValidatorToDelegation.IsSet())
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegationManager.go:498]: I would leave like this

File: vm/systemSmartContracts/validator.go (hunk 14 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go:1900]: done

File: vm/systemSmartContracts/delegation.go [hunk 29]
```diff
@@ -301,7 +324,165 @@ func (d *delegation) init(args *vmcommon.ContractCallInput) vmcommon.ReturnCode
...
+	returnCode = d.delegateUser(validatorData.TotalStakeValue, ownerAddress, args.RecipientAddr, dStatus)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	return vmcommon.Ok
+}
+
+func (d *delegation) getBLSKeyStatus(key []byte) (uint32, error) {
+	data := d.eei.GetStorageFromAddress(d.stakingSCAddr, key)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:446]: it is a little bit different as I do not initiate the whole structure and will return a status. yes it can be integrated however it brings a lot of changes.

File: vm/systemSmartContracts/validator.go (hunk 11 above)
Comment (iulianpascalau) [reply vm/systemSmartContracts/validator.go:1927]: right

File: vm/systemSmartContracts/staking.go (hunk 15 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go]: done

File: vm/systemSmartContracts/staking.go [hunk 30]
```diff
@@ -1744,7 +1744,7 @@ func (s *stakingSC) changeOwnerAndRewardAddress(args *vmcommon.ContractCallInput
 		}
 
 		if len(registrationData.RewardAddress) == 0 {
-			s.eei.AddReturnMessage("cannot unStake a key that is not registered")
+			s.eei.AddReturnMessage("cannot change owner and reward address a key is not registered")
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/staking.go:1747]: for a key which is not registered

Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:1747]: done

**Review by SebastianMarian — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]
